File: CodeChef/Long/September 2017/SUMCUBE.py
from itertools import combinations
import sys
sys.setrecursionlimit(1000000)

mod = 1000000007
powTwo = [1]
for _ in range(100005):
    powTwo.append((powTwo[-1]*2)%mod)

def setVariables1():
    global OneEdgeUsingTwoNodes
    global OneEdgeUsingTwoNodesAns
    OneEdgeUsingTwoNodes = e
    OneEdgeUsingTwoNodesAns = (powTwo[v-2]*OneEdgeUsingTwoNodes)%mod

def setVariables2():
    global TwoEdgeUsingThreeNodes,TwoEdgeUsingFourNodes
    global TwoEdgeUsingThreeNodesAns,TwoEdgeUsingFourNodesAns
    for i in edgeFrom[1:]:
        if(i>=2):
            TwoEdgeUsingThreeNodes += (i*(i-1))/2
    TwoEdgeUsingThreeNodesAns = powTwo[v-3]*TwoEdgeUsingThreeNodes
    if v>=4:
        TwoEdgeUsingFourNodes = (e*(e-1))/2 - TwoEdgeUsingThreeNodes
        TwoEdgeUsingFourNodesAns = powTwo[v-4]*TwoEdgeUsingFourNodes

def setVariables3():
    global ThreeEdgeUsingThreeNodes,ThreeEdgeUsingFourNodes,ThreeEdgeUsingFiveNodes,ThreeEdgeUsingSixNodes
    global ThreeEdgeUsingThreeNodesAns,ThreeEdgeUsingFourNodesAns,ThreeEdgeUsingFiveNodesAns,ThreeEdgeUsingSixNodesAns
    for a in range(1,v+1):
        edgesFromA = edgeFrom[a]
        # For 4 Nodes
        if(edgesFromA >= 3):
            ThreeEdgeUsingFourNodes += (edgesFromA*(edgesFromA-1)*(edgesFromA-2))/6
        # Dependency on connected nodes b and c
        for i in range(edgesFromA):
            b = graph[a][i]
            for j in range(i+1,edgesFromA):
                c = graph[a][j]
                # For 3 Nodes
                if((b,c) in graphDict):
                    ThreeEdgeUsingThreeNodes += 1
                    ThreeEdgeUsingFiveNodes += 1    # Because of connection between b&c
                # For 5 Nodes
                ThreeEdgeUsingFiveNodes += e-edgeFrom[a]-edgeFrom[b]-edgeFrom[c]+2
                ThreeEdgeUsingFiveNodes %= mod
    ThreeEdgeUsingThreeNodes /= 3
    # For 4 Nodes Straight Lines
    temp = 0
    for a in range(1,v+1):
        edgesFromA = edgeFrom[a]
        for i in range(edgesFromA):
            b = graph[a][i]
            for j in range(edgeFrom[b]):
                c = graph[b][j]
                if c == a:
                    continue
                temp += edgeFrom[c]-1
                if (a,c) in graphDict:
                    temp -= 1
    ThreeEdgeUsingFourNodes += temp/2
    ThreeEdgeUsingFourNodes %= mod
    ThreeEdgeUsingSixNodes = (((e*(e-1)*(e-2))/6)%mod - (ThreeEdgeUsingThreeNodes+ThreeEdgeUsingFourNodes+ThreeEdgeUsingFiveNodes)%mod)%mod
    # print ThreeEdgeUsingThreeNodes,ThreeEdgeUsingFourNodes,ThreeEdgeUsingFiveNodes,ThreeEdgeUsingSixNodes
    if v>=3:
        ThreeEdgeUsingThreeNodesAns = (6*powTwo[v-3]*ThreeEdgeUsingThreeNodes)%mod
    if v>=4:
        ThreeEdgeUsingFourNodesAns = (6*powTwo[v-4]*ThreeEdgeUsingFourNodes)%mod
    if v>=5:
        ThreeEdgeUsingFiveNodesAns = (6*powTwo[v-5]*ThreeEdgeUsingFiveNodes)%mod
    if v>=6:
        ThreeEdgeUsingSixNodesAns = (6*powTwo[v-6]*ThreeEdgeUsingSixNodes)%mod

for _ in range(input()):
    v,e,k = map(int,raw_input().split())
    graph = [[] for __ in range(v+1)]
    edgeFrom = [0 for i in range(v+1)]
    graphDict = {}
    for __ in range(e):
        a,b = map(int,raw_input().split())
        graph[a].append(b)
        graph[b].append(a)
        edgeFrom[a] += 1
        edgeFrom[b] += 1
        graphDict[(a,b)] = True
        graphDict[(b,a)] = True
    if v == 1 or e == 0:
        print 0
    elif v == 2:
        print 1
    else:
        # One Edge in expansion
        OneEdgeUsingTwoNodes = 0
        # Two Edges in expansion
            # using 3 nodes
        TwoEdgeUsingThreeNodes = 0
        TwoEdgeUsingThreeNodesAns = 0
            # using 4 nodes
        TwoEdgeUsingFourNodes = 0
        TwoEdgeUsingFourNodesAns = 0
        # Three Edges in expansion
            # using 3 nodes
        ThreeEdgeUsingThreeNodes = 0
        ThreeEdgeUsingThreeNodesAns = 0
            # using 4 nodes
        ThreeEdgeUsingFourNodes = 0
        ThreeEdgeUsingFourNodesAns = 0
            # using 5 nodes
        ThreeEdgeUsingFiveNodes = 0
        ThreeEdgeUsingFiveNodesAns = 0
            # using 6 nodes
        ThreeEdgeUsingSixNodes = 0
        ThreeEdgeUsingSixNodesAns = 0

        if k == 1:
            setVariables1()
            print (OneEdgeUsingTwoNodesAns)%mod
        elif k == 2:
            setVariables1()
            setVariables2()
            TwoEdgeUsingThreeNodesAns = (2*TwoEdgeUsingThreeNodesAns)%mod
            TwoEdgeUsingFourNodesAns = (2*TwoEdgeUsingFourNodesAns)%mod
            print (OneEdgeUsingTwoNodesAns+TwoEdgeUsingThreeNodesAns+TwoEdgeUsingFourNodesAns)%mod
        else:
            setVariables1()
            setVariables2()
            setVariables3()
            TwoEdgeUsingThreeNodesAns = (6*TwoEdgeUsingThreeNodesAns)%mod
            TwoEdgeUsingFourNodesAns = (6*TwoEdgeUsingFourNodesAns)%mod
            # print OneEdgeUsingTwoNodesAns,TwoEdgeUsingThreeNodesAns,TwoEdgeUsingFourNodesAns,ThreeEdgeUsingThreeNodesAns,ThreeEdgeUsingFourNodesAns,ThreeEdgeUsingFiveNodesAns,ThreeEdgeUsingSixNodesAns
            print (OneEdgeUsingTwoNodesAns+TwoEdgeUsingThreeNodesAns+TwoEdgeUsingFourNodesAns+ThreeEdgeUsingThreeNodesAns+ThreeEdgeUsingFourNodesAns+ThreeEdgeUsingFiveNodesAns+ThreeEdgeUsingSixNodesAns)%mod

File: CodeChef/Long/September 2017/SUMCUBE.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long
#define ull unsigned long long
#define rep(i,s,e) for(ll i=s; i<e; i++)
#define pb push_back
#define mp make_pair
#define mod 1000000007

ll powTwo[100005];
vector< vector<ll> > graph(100005);
ll edgeFrom[100005];
ll twoLevelEdgeFrom[100005];
unordered_map<ll,bool> graphDict[100005];
ll v,e,k;

// Variables
// One Edge in expansion
ll OneEdgeUsingTwoNodes = 0;
ll OneEdgeUsingTwoNodesAns = 0;
// Two Edges in expansion
    // using 3 nodes
ll TwoEdgeUsingThreeNodes = 0;
ll TwoEdgeUsingThreeNodesAns = 0;
    // using 4 nodes
ll TwoEdgeUsingFourNodes = 0;
ll TwoEdgeUsingFourNodesAns = 0;
// Three Edges in expansion
    // using 3 nodes
ll ThreeEdgeUsingThreeNodes = 0;
ll ThreeEdgeUsingThreeNodesAns = 0;
    // using 4 nodes
ll ThreeEdgeUsingFourNodes = 0;
ll ThreeEdgeUsingFourNodesAns = 0;
    // using 5 nodes
ll ThreeEdgeUsingFiveNodes = 0;
ll ThreeEdgeUsingFiveNodesAns = 0;
    // using 6 nodes
ll ThreeEdgeUsingSixNodes = 0;
ll ThreeEdgeUsingSixNodesAns = 0;
void resetVaribles(ll v) {
  rep(i,0,v+5) {
    graph[i] = vector<ll>();
    edgeFrom[i] = 0;
    graphDict[i] = unordered_map<ll,bool>();
  }
  OneEdgeUsingTwoNodes = 0;
  TwoEdgeUsingThreeNodes = 0;
  TwoEdgeUsingThreeNodesAns = 0;
  TwoEdgeUsingFourNodes = 0;
  TwoEdgeUsingFourNodesAns = 0;
  ThreeEdgeUsingThreeNodes = 0;
  ThreeEdgeUsingThreeNodesAns = 0;
  ThreeEdgeUsingFourNodes = 0;
  ThreeEdgeUsingFourNodesAns = 0;
  ThreeEdgeUsingFiveNodes = 0;
  ThreeEdgeUsingFiveNodesAns = 0;
  ThreeEdgeUsingSixNodes = 0;
  ThreeEdgeUsingSixNodesAns = 0;
}

void setVariables1() {
  OneEdgeUsingTwoNodes = e;
  OneEdgeUsingTwoNodesAns = (powTwo[v-2]*OneEdgeUsingTwoNodes)%mod;
}

void setVariables2() {
  rep(z,1,v+1) {
    ll i = edgeFrom[z];
    if(i==0) continue;
    else if(i>=2) TwoEdgeUsingThreeNodes += (i*(i-1))/2;
  }
  TwoEdgeUsingThreeNodesAns = (powTwo[v-3]*TwoEdgeUsingThreeNodes)%mod;
  if(v>=4) {
    TwoEdgeUsingFourNodes = (e*(e-1))/2 - TwoEdgeUsingThreeNodes;
    TwoEdgeUsingFourNodesAns = (powTwo[v-4]*TwoEdgeUsingFourNodes)%mod;
  }
}

bool compareForSort(pair<ll,ll> i1, pair<ll,ll> i2) {
    return (i1.second < i2.second);
}

void solveForTriangles() {
  pair<ll,ll> vAndE[v];
  rep(i,1,v+1) {
    vAndE[i-1] = mp(i,edgeFrom[i]);
  }
  sort(vAndE,vAndE+v,compareForSort);
  bool isNodePresent[v+1];
  rep(i,0,v+1) {
    isNodePresent[i] = true;
  }
  for(auto z: vAndE) {
    ll a = z.first;
    ll edgesFromA = edgeFrom[a];
    rep(i,0,edgesFromA) {
      ll b = graph[a][i];
      if(isNodePresent[b]) {
        rep(j,i+1,edgesFromA) {
          ll c = graph[a][j];
          if(isNodePresent[c] && graphDict[b].find(c) != graphDict[b].end()) {
            ThreeEdgeUsingThreeNodes += 1;
          }
        }
      }
    }
    isNodePresent[a] = false;
  }
}

void setVariables3() {
  // For 3 Nodes
  solveForTriangles();
  // For 4 Nodes
  ll temp1=0,temp2 = 0;
  ll count2 = 0;
  rep(a,1,v+1) {
    ll edgesFromA = edgeFrom[a];
    if(edgesFromA >= 3) {
      temp1 += (edgesFromA*(edgesFromA-1)*(edgesFromA-2))/6;
      temp1 %= mod;
    }
    rep(i,0,edgesFromA) {
      ll b = graph[a][i];
      temp2 += twoLevelEdgeFrom[b]-edgesFromA+1;
    }
  }
  temp2 >>= 1;
  temp2 -= 3*ThreeEdgeUsingThreeNodes; // Because of connection between b&c
  ThreeEdgeUsingFourNodes = temp1 + temp2;
  ThreeEdgeUsingFourNodes %= mod;
  // For 5 Nodes
  rep(a,1,v+1) {
    ll twoLevelEdge = twoLevelEdgeFrom[a];
    ThreeEdgeUsingFiveNodes += twoLevelEdge;
  }
  ThreeEdgeUsingFiveNodes /= 2;
  ThreeEdgeUsingFiveNodes *= (e-2);
  ThreeEdgeUsingFiveNodes -= 3*temp1;
  ThreeEdgeUsingFiveNodes -= 2*temp2;
  ThreeEdgeUsingFiveNodes -= 3*ThreeEdgeUsingThreeNodes;
  ThreeEdgeUsingFiveNodes += mod;
  ThreeEdgeUsingFiveNodes %= mod;
  // For 6 Nodes
  // printf("%lld %lld\n", ((e*(e-1)*(e-2))/6)%mod, (ThreeEdgeUsingThreeNodes+ThreeEdgeUsingFourNodes+ThreeEdgeUsingFiveNodes)%mod);
  ThreeEdgeUsingSixNodes = ((((e*(e-1)*(e-2))/6)%mod) - ((ThreeEdgeUsingThreeNodes+ThreeEdgeUsingFourNodes+ThreeEdgeUsingFiveNodes)%mod) + mod)%mod;
  // printf("%lld %lld %lld %lld\n", ThreeEdgeUsingThreeNodes,ThreeEdgeUsingFourNodes,ThreeEdgeUsingFiveNodes,ThreeEdgeUsingSixNodes);
  // Calculating answers
  if(v>=3) ThreeEdgeUsingThreeNodesAns = (6*(powTwo[v-3]*ThreeEdgeUsingThreeNodes)%mod)%mod;
  if(v>=4) ThreeEdgeUsingFourNodesAns = (6*(powTwo[v-4]*ThreeEdgeUsingFourNodes)%mod)%mod;
  if(v>=5) ThreeEdgeUsingFiveNodesAns = (6*(powTwo[v-5]*ThreeEdgeUsingFiveNodes)%mod)%mod;
  if(v>=6) ThreeEdgeUsingSixNodesAns = (6*(powTwo[v-6]*ThreeEdgeUsingSixNodes)%mod)%mod;
}

int main() {
  powTwo[0] = 1;
  rep(i,1,100005) {
    powTwo[i] = (powTwo[i-1]*2)%mod;
  }
  ll t;
  scanf("%lld\n", &t);
  rep(_,0,t) {
    scanf("%lld %lld %lld\n",&v,&e,&k);
    resetVaribles(v);
    rep(__,0,e) {
      ll a,b;
      scanf("%lld %lld\n",&a,&b);
      graph[a].pb(b);
      graph[b].pb(a);
      edgeFrom[a] += 1;
      edgeFrom[b] += 1;
      graphDict[a][b] = true;
      graphDict[b][a] = true;
    }

    rep(a,1,v+1) {
      ll edgesFromA = edgeFrom[a];
      ll temp = 0;
      rep(i,0,edgesFromA) {
        ll b = graph[a][i];
        temp += edgeFrom[b]-1;
      }
      twoLevelEdgeFrom[a] = temp;
    }

    if(v==1 || e==0) {
      printf("%d\n",0);
    }
    else if(v==2) {
      printf("%d\n",1);
    }
    else {
      if(k == 1) {
        setVariables1();
        printf("%lld\n",(OneEdgeUsingTwoNodesAns)%mod);
      }
      else if(k == 2) {
        setVariables1();
        setVariables2();
        TwoEdgeUsingThreeNodesAns = (2*TwoEdgeUsingThreeNodesAns)%mod;
        TwoEdgeUsingFourNodesAns = (2*TwoEdgeUsingFourNodesAns)%mod;
        ll finalAns = (OneEdgeUsingTwoNodesAns+TwoEdgeUsingThreeNodesAns+TwoEdgeUsingFourNodesAns)%mod;
        printf("%lld\n", finalAns);
      }
      else {
        setVariables1();
        setVariables2();
        setVariables3();
        TwoEdgeUsingThreeNodesAns = (6*TwoEdgeUsingThreeNodesAns)%mod;
        TwoEdgeUsingFourNodesAns = (6*TwoEdgeUsingFourNodesAns)%mod;
        // printf("%lld %lld %lld %lld %lld %lld %lld\n",OneEdgeUsingTwoNodesAns,TwoEdgeUsingThreeNodesAns,TwoEdgeUsingFourNodesAns,ThreeEdgeUsingThreeNodesAns,ThreeEdgeUsingFourNodesAns,ThreeEdgeUsingFiveNodesAns,ThreeEdgeUsingSixNodesAns);
        ll finalAns = (OneEdgeUsingTwoNodesAns+TwoEdgeUsingThreeNodesAns+TwoEdgeUsingFourNodesAns+ThreeEdgeUsingThreeNodesAns+ThreeEdgeUsingFourNodesAns+ThreeEdgeUsingFiveNodesAns+ThreeEdgeUsingSixNodesAns)%mod;
        printf("%lld\n", finalAns);
      }
    }
  }
  return 0;
}
